File: 0904-fruit-into-baskets/0904-fruit-into-baskets.cpp
class Solution {
public:
    int totalFruit(vector<int>& tree) {
         int i = 0, j = 0;
    int ans = 0;
    int count = 0;
    unordered_map<int, int> mp;
    while (j < tree.size())
    {
        mp[tree[j]]++;
        if (mp[tree[j]] == 1)
        {
            count++;
        }
        if (count > 2)
        {
            while (mp[tree[i]] > 1)
            {
                mp[tree[i]]--;
                i++;
            }
            mp[tree[i]]--;
            i++;
            count--;
        }
        ans = max(ans, j - i + 1);
        j++;
    }
    return ans;
    }
};

File: 0904-fruit-into-baskets/0904-fruit-into-baskets.java
class Solution {
    public int totalFruit(int[] fruits) {
         int lastFruit = -1;
        int secondlastFruit = -1;
        int currentMax = 0;
        int maxi = 0;
        int last_fruit_count = 0;
        for (Integer nums : fruits)
        {
            if (nums == lastFruit || nums == secondlastFruit)
            {
                currentMax += 1;
            }
            else
            {
                currentMax = last_fruit_count + 1;
            }
            if (nums == lastFruit)
            {
               last_fruit_count += 1;
            }
            else
            {
                last_fruit_count = 1;
            }
            if (nums != lastFruit)
            {
                secondlastFruit = lastFruit;
                lastFruit = nums;
            }
            maxi = Math.max(currentMax, maxi);
        }
        return maxi; 
    }
}